[WFLY-11751] Error in JBoss Management Console when clicking on the deployed subsystem jaxrs/rest-resource

gaol opened the pull request:
Issue: https://issues.jboss.org/browse/WFLY-11751

Changes in the PR:
- Move the code of adding management model from `POST_MODULE(JaxrsScanningProcessor.java)` to `INSTALL(JaxrsIntegrationProcessor.java)`


~~Recording scanned rest resource classes into a set as an DeploymentUnit's Attachment so that they can be used for adding the management resources under jaxrs subsystem.~~
~~A new attachment is used instead of using the existed `ADDITIONAL_RESTEASY_DEPLOYMENT_DATA`  because if the scanned resource classes are EJBs as well, they will be removed from `ResteasyDeploymentData`~~


- In case of EAR deployment, even the REST endpoints are defined in the ejb module, they can only be read out under the war sub-deployment.
   -  Using `/deployment=test.ear/subdeployment=web.war/subsystem=jaxrs/rest-resource=xx:read-resource(include-runtime)`

- In case of multiple wars inside one ear deployment, the `reset-resource` of one war will only display the  resource classes which are accessible to the war.

Please make sure your PR meets the following requirements:
- [ ] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue(s)
- [ ] Pull Request contains description of the issue(s)
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted

gaol committed:
[WFLY-11751] Error in JBoss Management Console when clicking on the deployed subsystem jaxrs/rest-resource

kabir commented:
@asoldano @ronsigal please approve

marekkopecky commented:
> In case of EAR deployment, even the REST endpoints are defined in the ejb module, they can only be read out under the war sub-deployment. 

@gaol  This approach could be confusing. What if ear file have two or more war? In which war the end-point is showed? Both? First-one?

I'm trying  kitchensink-ear example described on the jira. End-point in web archive doesn't work (no regression, but still wrong):
```
[standalone@localhost:9990 /] /deployment=kitchensink-ear.ear/subdeployment=kitchensink-ear-web.war/subsystem=jaxrs/rest-resource=org.jboss.as.quickstarts.kitchensink_ear.rest.MemberResourceRESTService:read-resource(include-runtime)
{
    "outcome" => "failed",
    "rolled-back" => true
}
[standalone@localhost:9990 /]
```
```
[standalone@localhost:9990 /] /deployment=kitchensink-ear.ear/subdeployment=kitchensink-ear-web.war/subsystem=jaxrs:read-resource(include-runtime,recursive)
{
    "outcome" => "failed",
    "rolled-back" => true
}
[standalone@localhost:9990 /] 
```
```
14:18:16,863 ERROR [org.jboss.as.controller.management-operation] (management-handler-thread - 1) WFLYCTL0013: Operation ("read-attribute") failed - address: ([
    ("deployment" => "kitchensink-ear.ear"),
    ("subdeployment" => "kitchensink-ear-web.war"),
    ("subsystem" => "jaxrs"),
    ("rest-resource" => "org.jboss.as.quickstarts.kitchensink_ear.rest.MemberResourceRESTService")
]) - failure description: "org.jboss.as.quickstarts.kitchensink_ear.rest.MemberResourceRESTService from [Module \"deployment.kitchensink-ear.ear.kitchensink-ear-ejb.jar\" from Service Module Loader]"
```

gaol commented:
@marekkopecky I tried [latest kitchensink-ear](https://github.com/wildfly/quickstart/tree/master/kitchensink-ear) following similar changes described in the Jira to move the REST endpoints from web to ejb module against my branch, and it works. So would you please double check whether you are using the correct branch or is the same error?

> This approach could be confusing. What if ear file have two or more war? In which war the end-point is showed? Both? First-one?

Good catch Marek !

In principal, I think as long as there are REST activator defined in the web module, and there are valid REST URIs to trigger the defined REST endpoints starts with that web module's context path, there should be `rest-resource` resources displayed under that web module

Currently the `rest-resource` resources will be displayed even there are no REST activator defined in the web module, I will update the PR to fix that.

marekkopecky commented:
> I tried latest kitchensink-ear ...

I use master branch as well. I still see an error described in my previous comment.

> In principal, I think as long as there are REST activator defined in the web module

EJB module contains its own REST activator in the example from jira. But this activator is available through kitchensink-ear-web module:
* http://localhost:8080/kitchensink-ear-web/rest2/members2
* http://localhost:8080/kitchensink-ear-web/rest/members

 @ronsigal  WDYT about this approach?

gaol commented:
@marekkopecky @ronsigal  I have updated the PR, the PR description above is still valid.

I made a reproducer at: https://github.com/gaol/quickstart/tree/WFLY-11751-reproducer/kitchensink-ear, and it passed the tests. @marekkopecky would you please take a look at the reproducer to see if there is anything that I missed? Thank you :) 

FYI: You can refer the the [WFLY-11751-reproducer.adoc](https://github.com/gaol/quickstart/blob/WFLY-11751-reproducer/kitchensink-ear/WFLY-11751-reproducer.adoc) on how to use the reproducer

marekkopecky commented:
@gaol Thank you for your readme file and for your quickstart repo example. But jira quickstart scenario is different. According to the first comment on the jira, @claudio4j copied both JaxRsActivator (rest -> rest2) and MemberResourceRESTService (members -> members2) to ejb subdeployment. Both:
* http://localhost:8080/kitchensink-ear-web/rest/members
* http://localhost:8080/kitchensink-ear-web/rest2/members2
End point works correctly, but `/deployment=kitchensink-ear.ear/subdeployment=kitchensink-ear-web.war/subsystem=jaxrs:read-resource(include-runtime,recursive)` still doesn't work.

marekkopecky commented:
@gaol Looking to your example. CLI works correctly in this example, but web-console is confusing:
1. start web-console
2. check web3 deployment: no rest end-point is shown
3. check web2, there is 2 end-pint
4. check web3 again, there is 2 end-point now
5. click to one end-point in web3, this click redirects to web2

I refer @zebrik for web-console issue ...

gaol commented:
@marekkopecky, I updated the reproducer by adding a `JaxRsActivator.java` to the ejb module with `@ApplicationPath("/rest0")` defined. And I can see the REST endpoints displayed from the `web3` sub deployment where no java classes defined in it.

marekkopecky commented:
@ronsigal WDYT about this PR? Can you please ensure that original use-cases from the jira are fix by this PR?

ronsigal commented:
Hi @marekkopecky,

Kabir also asked me to look at this. As soon as 4.0.0.CR1 goes out the door, I will. Assume it will take a little time, since I haven't worked on the management model stuff before. Looking forward to it, though. ;-)

-Ron

bstansberry commented:
Is there going to be any difference in the output for a complex case that wasn't failing before? Looking at the code I have to think too hard to figure that out. ;) . 

Re Marek's "What if ear file have two or more war? In which war the end-point is showed? Both? First-one?" question before, the resolution wasn't clear to me. I understand we want to report info for multiple wars if they both are relevant; I'm not clear though on whether what we are reporting for each is correct or whether it would not include things that are not relevant for a particular war.

gaol commented:
@bstansberry There is no difference in the output. :)

For the `multiple wars in ear` case, it has problem indeed. It omits the dependency definitions between each other within the EAR deployment. It will lead to non-relevant REST endpoints displayed under the war sub-deployment.

Basically, I think if the REST endpoint can be accessed via the URI:
`WEB-CONTEXT-PATH/APPLICATION_PATH/RESOURCE_PATH/RESOURCE_METHOD_PATH`
the REST endpoint information should be displayed under the web module associated with the `WEB-CONTEXT-PATH`, no matter in which module the class is defined, otherwise, it should not be displayed.

I will try to fix that.

gaol commented:
@marekkopecky @ronsigal @bstansberry I have update the PR and the description to reveal the changes. Could you please review ? Thanks :)

ronsigal commented:
I've just started reading http://docs.wildfly.org/16/Extending_WildFly.html .

bstansberry commented on jaxrs/src/main/java/org/jboss/as/jaxrs/DeploymentRestResourcesDefintion.java:
```diff
@@ -184,71 +187,88 @@ public void execute(OperationContext context, ModelNode operation) throws Operat
+                                                List<JaxrsResourceMethodDescription> resMethodInvokers = new ArrayList<>();
+                                                List<JaxrsResourceLocatorDescription> resLocatorInvokers = new ArrayList<>();
+                                                for (Map.Entry<String, List<ResourceInvoker>> resource : registry
+                                                        .getBounded().entrySet()) {
```
This would have been quite a bit easier to review if these line break changes weren't there.

A secret magic trick for looking at these is to add ?w=1 to the github URL:

https://github.com/wildfly/wildfly/pull/12138/files?w=1

If you do that github will display changes due to whitespace very differently; what seems like really large changes because blocks of code got indented to another level, like there are here, disappear.

Once you do that and look at the change in this file, a whole lot of the remaining changes are just due to these different line breaks. Eliminate those and this complex scary change suddenly is quite simple. You're looking up the deploymentModule so you can load from it, and you're processing a List<HttpServletDispatcher> resteasyServlets instead of Servlet resteasyServlet. And you throw ISE instead of OFE if a CNFE is caught.

I understand why the line breaks were added; indenting another level pushes a lot of code off the screen.

@ronsigal FYI, in case that makes it easier to see what this is doing.

gaol commented on jaxrs/src/main/java/org/jboss/as/jaxrs/DeploymentRestResourcesDefintion.java:
```diff
@@ -184,71 +187,88 @@ public void execute(OperationContext context, ModelNode operation) throws Operat
-
+                StringBuilder sb = new StringBuilder(ModelDescriptionConstants.DEPLOYMENT);
+                sb.append(".");
+                sb.append(address.getElement(0).getValue());
```
I am checking if it is possible to bypass the necessary to get current deployment module. and I will optimize the code. Thanks.

gaol commented:
@bstansberry @ronsigal I have updated PR considering cache some data in OperationContext.

Changes in summary:
- Cache the `UndertowDeploymentService` in `OperationContext`. This service is used to find the RESTEasy servlets(`HttpServletDispatcher`)
- Cache the deploymentModule in `OperationContext`. This Module is needed(I did not find a way to bypass the check) to load the resource class by the class name to check if the resource class is the implementation class in case the `@Path` is defined in the interface. 
- Loop for each RESTEasy servlets(`HttpServletDispatcher`) instead of picking the first one. It is possible to have multiple RESTEasy servlets(or multiple REST Application) in one deployment, it will affect the `resource-methods` which takes the `servlet-mappings` into consideration.
- Fixed the whitespace problem
- Cache the resource class related meta info so that it will be checked once only for multiple(3 in the case) attributes.

bstansberry commented:
@ronsigal This LGTM.

ronsigal commented:
Hey Brian,

I'm sorry I never got around to looking at this. In the past month or two I've read the Wildfly extension document (several times) and looked at the jaxrs code. Then I let it go for a while to focus on our next RESTEasy release in two weeks. If you like, I will still try to take a look. Of course, I trust your judgment. ;-)

-Ron
